### PR TITLE
Reload the client upon a rebalance

### PR DESCRIPTION
**Pull request by Tang8330**

### CODE REVIEW

File: lib/kafkalib/consumer.go [hunk 1]
```diff
@@ -10,3 +12,31 @@ type Consumer interface {
...
+func ShouldReload(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// [27] Rebalance In Progress: the coordinator has begun rebalancing the group, the client should rejoin the group
+	return errors.Is(err, kafka.RebalanceInProgress)
+}
+
+func (r *Reader) Reload() error {
```
Comment (nathan-artie) [inline lib/kafkalib/consumer.go:37]: This never returns an error.

File: processes/consumer/kafka.go [hunk 2]
```diff
@@ -101,12 +101,19 @@ func StartConsumer(ctx context.Context, cfg config.Config, inMemDB *models.Datab
...
 			for {
 				kafkaMsg, err := kafkaConsumer.FetchMessage(ctx)
 				if err != nil {
+					if kafkalib.ShouldReload(err) {
+						if err = kafkaConsumer.Reload(); err != nil {
+							logger.Fatal("Failed to reload kafka consumer", slog.Any("err", err))
+						}
+					}
+
 					slog.With(artie.KafkaMsgLogFields(kafkaMsg)...).Warn("Failed to read kafka message", slog.Any("err", err))
```
Comment (nathan-artie) [inline processes/consumer/kafka.go:115]: Put this in an `else` since if we reload we don't want to log?

File: processes/consumer/kafka.go [hunk 3]
```diff
@@ -101,12 +101,19 @@ func StartConsumer(ctx context.Context, cfg config.Config, inMemDB *models.Datab
...
 			}
 
-			kafkaConsumer := kafka.NewReader(kafkaCfg)
+			kafkaConsumer := kafkalib.NewReader(kafkaCfg)
 			topicToConsumer.Add(topic, kafkaConsumer)
 			for {
 				kafkaMsg, err := kafkaConsumer.FetchMessage(ctx)
 				if err != nil {
+					if kafkalib.ShouldReload(err) {
+						if err = kafkaConsumer.Reload(); err != nil {
```
Comment (nathan-artie) [inline processes/consumer/kafka.go]: Might want to log here that we're going to reload.

File: lib/kafkalib/consumer.go (hunk 1 above)
Comment (Tang8330) [reply lib/kafkalib/consumer.go]: thanks

File: processes/consumer/kafka.go (hunk 3 above)
Comment (Tang8330) [reply processes/consumer/kafka.go]: yeah just added

**Review by nathan-artie — APPROVED** [review]

File: lib/kafkalib/consumer.go [hunk 4]
```diff
@@ -10,3 +12,31 @@ type Consumer interface {
...
+
+func NewReader(config kafka.ReaderConfig) *Reader {
+	return &Reader{
+		Reader: kafka.NewReader(config),
+		config: config,
+	}
+}
+
+func ShouldReload(err error) bool {
+	if err == nil {
```
Comment (nathan-artie) [inline lib/kafkalib/consumer.go:29]: This could go away, `errors.Is` will return `false` for `nil`.

Comment (nathan-artie) [reply lib/kafkalib/consumer.go:29]: https://go.dev/play/p/547vdvNWgJQ

Comment (Tang8330) [reply lib/kafkalib/consumer.go:29]: ill do that tomorrow and also add a test

**Pull request merged**